Brokerages/Fxcm/FxcmBrokerage.cs

```diff
@@ -0,0 +1,364 @@
+                    if (_orderEventQueue.IsEmpty || !_orderEventQueue.TryDequeue(out orderEvent))
+                    {
+                        Thread.Sleep(1);
+                        continue;
+                    }
+
+                    OnOrderEvent(orderEvent);
+                }
+            });
+            _orderEventThread.Start();
+            while (!_orderEventThread.IsAlive) {}
+
+            // create the gateway
+            _gateway = GatewayFactory.createGateway();
+
+            // register the message listeners with the gateway
+            _gateway.registerGenericMessageListener(this);
+            _gateway.registerStatusMessageListener(this);
+
+            // create local login properties
+            var loginProperties = new FXCMLoginProperties(_userName, _password, _terminal, _server);
+
+            // log in
+            _gateway.login(loginProperties);
+
+            // load instruments, accounts, orders, positions
+            LoadInstruments();
+            LoadAccounts();
+            LoadOpenOrders();
+            LoadOpenPositions();
+        }
+
+        /// <summary>
+        /// Disconnects the client from the broker's remote servers
+        /// </summary>
+        public override void Disconnect()
+        {
+            if (!IsConnected) return;
+
+            Log.TraceFormat("FxcmBrokerage.Disconnect()");
+
+            // log out
+            _gateway.logout();
+
+            // remove the message listeners
+            _gateway.removeGenericMessageListener(this);
+            _gateway.removeStatusMessageListener(this);
+
+            // request and wait for thread to stop
+            _orderEventThreadStopped = true;
+            _orderEventThread.Join();
+        }
+
+        /// <summary>
+        /// Gets all open orders on the account. 
+        /// NOTE: The order objects returned do not have QC order IDs.
+        /// </summary>
+        /// <returns>The open orders returned from FXCM</returns>
+        public override List<Order> GetOpenOrders()
+        {
+            Log.TraceFormat("FxcmBrokerage.GetOpenOrders()");
+
+            return _openOrders.Values.ToList().Where(x => OrderIsOpen(x.getFXCMOrdStatus().getCode())).Select(ConvertOrder).ToList();
+        }
+
+        /// <summary>
+        /// Gets all holdings for the account
+        /// </summary>
+        /// <returns>The current holdings from the account</returns>
+        public override List<Holding> GetAccountHoldings()
+        {
+            Log.TraceFormat("FxcmBrokerage.GetAccountHoldings()");
+
+            var holdings = _openPositions.Values.Select(ConvertHolding).ToList();
+
+            // Set MarketPrice in each Holding
+            var symbols = holdings.Select(x => ConvertSymbolToFxcmSymbol(x.Symbol)).ToList();
+            if (symbols.Count > 0)
+            {
+                var quotes = GetQuotes(symbols).ToDictionary(x => x.getInstrument().getSymbol());
+                foreach (var holding in holdings)
```

Comment:
IB has a tendency to send back 0 quantity holdings, maybe we should check for something similar here as well to prevent subscribing to data we don't need